GEOSlandassim_GridComp/clsm_ensupd_read_obs.F90

```diff
@@ -2145,7 +2145,511 @@ subroutine read_obs_sm_ASCAT_EUMET(                                         &
+       err_msg = 'CYGNSS SM obs file not found!'
+       call ldas_abort(LDAS_GENERIC_ERROR, Iam, err_msg)
+    end if
+
+    ! Open the NetCDF observation file
+    ierr = nf90_open(trim(tmpfname), nf90_nowrite, ncid)
+
+    ! get variable dimension IDs
+    ierr = nf90_inq_dimid(ncid, 'lon',        lon_dimid)
+    ierr = nf90_inq_dimid(ncid, 'lat',        lat_dimid)
+    ierr = nf90_inq_dimid(ncid, 'time',       time_dimid)    
+    ierr = nf90_inq_dimid(ncid, 'timeslices', timeslices_dimid)
+    ierr = nf90_inq_dimid(ncid, 'startstop',  startstop_dimid)
+
+    ! dimensions sizes
+    ierr = nf90_inquire_dimension(ncid, lon_dimid,        len=N_lon)
+    ierr = nf90_inquire_dimension(ncid, lat_dimid,        len=N_lat)
+    ierr = nf90_inquire_dimension(ncid, time_dimid,       len=N_time)
+    ierr = nf90_inquire_dimension(ncid, timeslices_dimid, len=N_timeslices)
+    ierr = nf90_inquire_dimension(ncid, startstop_dimid,  len=N_startstop)
+
+    ! get variable IDs
+    ierr = nf90_inq_varid(ncid, 'SM_daily',       sm_d_varid)
+    ierr = nf90_inq_varid(ncid, 'SM_subdaily',    sm_subd_varid)
+    ierr = nf90_inq_varid(ncid, 'SIGMA_daily',    sigma_d_varid)
+    ierr = nf90_inq_varid(ncid, 'SIGMA_subdaily', sigma_subd_varid)
+    ierr = nf90_inq_varid(ncid, 'timeintervals',  timeintervals_varid)
+    ierr = nf90_inq_varid(ncid, 'latitude',       lat_varid)
+    ierr = nf90_inq_varid(ncid, 'longitude',      lon_varid)
+
+    ! allocate memory for the variables
+    allocate(sm_d(         N_lon, N_lat, N_time       ))
+    allocate(sm_subd(      N_lon, N_lat, N_timeslices ))
+    allocate(sigma_d(      N_lon, N_lat, N_time       ))
+    allocate(sigma_subd(   N_lon, N_lat, N_timeslices ))
+    allocate(timeintervals(N_startstop,  N_timeslices ))
+    allocate(latitudes(    N_lon, N_lat               ))
+    allocate(longitudes(   N_lon, N_lat               ))
+    allocate(tmp_sm(       N_lon, N_lat               ))
+    allocate(tmp_sigma(    N_lon, N_lat               ))
+
+    ! read the variables
+    ierr = nf90_get_var(ncid, timeintervals_varid, timeintervals)
+    ierr = nf90_get_var(ncid, lat_varid, latitudes)
+    ierr = nf90_get_var(ncid, lon_varid, longitudes)
+
+    ! read either subdaily or daily soil moisture and sigma variables
+    if ( trim(this_obs_param%descr) == 'CYGNSS_SM' ) then
+
+       ! subdaily observations required
+       ierr = nf90_get_var(ncid, sm_subd_varid,    sm_subd)
+       ierr = nf90_get_var(ncid, sigma_subd_varid, sigma_subd)
+
+       ! find the time slice that corresponds to the assimilation window
+       idx = -1
+       do i = 1, N_timeslices
+           if ((date_time%hour ==  3 .and. timeintervals(1,i) == 0.0 ) .or.            &
+               (date_time%hour ==  9 .and. timeintervals(1,i) == 0.25) .or.            &
+               (date_time%hour == 15 .and. timeintervals(1,i) == 0.5 ) .or.            &
+               (date_time%hour == 21 .and. timeintervals(1,i) == 0.75)         ) then
```

Comment:
This seems incompatible with assimilation windows that are centered on 3z, 6z, ... (as in [1:30,4:30], [4:30,7:30], ...)
Also, wouldn't it be easier to identify the idx value from H[xx], as in:  
```
if     H03 then
   idx=1
elseif H09 then
   idx=2
...
```